### DISCUSSION TITLE
Upgrade to TypeScript

### DISCUSSION THREAD

**Discussion by Ascor8522** [ScratchAddons/ScratchAddons#1301] [Discussion]
Please consider using [TypeScript](https://www.typescriptlang.org/) for the extension.
There are lots of benefits, including:
* the codebase is already ready
    * TypeScript is a superset of JavaScript so JavaScript code is valid TypeScript
    * just change the extension of all the files are you're good to go
    * TypeScript has better OO concepts, making it easier to deal with classes and inheritance
* it adds static checking, so it's less errors prone
    * typings are optionnal but recommended
    * you can even use typings from other libraries in your editor (Vue, DOM api, etc.)
* you can use modern JavaScript features, and TypeScript will transpile the code for older browsers
* it can generate typings files easily, to have an easier time using the API
* all code can be bundled into one file, making it faster to load

[See more reasons for using TypeScript](https://www.typescriptlang.org/why-create-typescript)

This doesn't requires lots of changes.
Also, TypeScript is easy to learn when you already know JavaScript.
Finally, the language is now used in major web /node projects.

I've entierly switched from using JavaScript to TypeScript and have never been happier. If you need more infos about TypeScript / have any questions, hit me up. I'd be glad to help with that.

**Comment by cobaltt7** [comment]
Yes but then it would have to be built and couldn't be used out of the box
________________________________

**Reply by Ascor8522** [reply]
Would anybody ever use this out of the box?
The general users would donwload it from the webstore out of simplicity and the "load unpack extension" is prmairly meant for developers.
CD/CI could be setup pretty easily, producing builds if needed.

Even if the only goal is to make it easier for testing, you still can compile and build the extension in watch mode.

But anway, if not having to be built is one of the core principles of the extension, I respect that.

**Reply by Hans5958** [reply]
Note: Most of the developers develops by just taking this repository zipball, do the necessary changes, and do the trouble of copying and pasting the changed contents. They don't even know how to use Git or Visual Studio Code.

I am one of them who uses Git and VSCode, and have properly set up the necessary tools to develop programs with multiple languages. I have experience on facing TS code and I know how easy it is. I am kind of disappointed about people didn't want to do this, but that's the truth that I and some others need to face.

**Reply by Ascor8522** [reply]
> Most of the developers develops by just taking this repository zipball, do the necessary changes, and do the trouble of copying and pasting the changed contents. They don't even know how to use Git or Visual Studio Code.

Can we even consider those people developers then? ^^'
It's often that same question that comes back.
How low should the entry point be, so that most people can contribute to the project.

Keep in mind people with a limited knowledge of programming won't add important / complicated features.
So making the code that accessible isn't benefitting anyone, nor the unexperienced nor the experienced developers.

> I am kind of disappointed about people didn't want to do this, but that's the truth that I and some others need to face.

Can't agree more on that

**Comment by pufferfish101007** [comment]
> you can use modern JavaScript features, and TypeScript will transpile the code for older browsers

We don't support older browsers anyway, as scratch doesn't support them either.

**Comment by apple502j** [comment]
Based on your misunderstandings on how TypeScript works, I am going to say no.

- You have confused three things: transpiling, polyfilling and minifying. TypeScript to JavaScript is transpiling, adding core-js is polyfilling, and merging them into one file and compressing is minifying.
- There is no need for polyfilling. We only support two modern browsers that implements most if not all of ES2020 features. No need to add es5-shim (required for Internet Explorer 8?) or core-js (required for Internet Explorer and old Edge) via babel. (Babel is NOT TypeScript!)
- Minifying does not make things faster to load. Extensions are loaded from your computer, so the bottleneck only exists for those who use old computers. Anyway this is a big no - stores don't like minified code.
- In terms of class and inheritance, vanilla features are just enough. Unless you missed ES6 class syntax, of course.
- **You can't make use of TypeScript if you did not provide type information.** TypeScript requires YOU to define types, associate them and correctly describe them in all of the function definitions. If you just rename the files to `.ts` you just complicate stuff without any profits.

tbh this really looks like an ad

**Reply by WorldLanguages** [reply]
Locked to avoid off-topic discussion about Typescript.
If anyone wants to discuss this, feel free to open an issue about adding build steps to Scratch Addons. Reality is tho, no building is one of the core ideas of Scratch Addons I have considered when starting the project.

**Reply by Hans5958** [reply]
@WorldLanguages Eh, I don't see any off-topic content that justifies this post to be locked. I saw this as a friendly discussion for the change. Reopening.

**Reply by Hans5958** [reply]
Technically the point of **Type**Script is the support of **types**, which is good if you want to avoid amateur bugs in the future.

**Reply by Ascor8522** [reply]
> Based on your misunderstandings on how TypeScript works, I am going to say no.

Based on your answer, I can see **you don't know any of the functionalities TypeScript offers**, **even more if they aren't TS's main features**. The **main goal was to show everything the language was capable of** and to show there are very useful built-in tools too.

> You have confused three things: transpiling, polyfilling and minifying. TypeScript to JavaScript is transpiling, adding core-js is polyfilling, and merging them into one file and compressing is minifying.

I have not confused those things. I mentionned those features because they might become quite useful at some point. But, please note **all of them are available as 1st party tools with TS**. There are no other dependencies, additional packages and setup needed. It's one package, done by the maintainers of the language, with the language and its usage in mind.

> There is no need for polyfilling. We only support two modern browsers that implements most if not all of ES2020 features.

Fair enough.

> Babel is NOT TypeScript!

I never pretended it was. **Babel**, let it be the only tool of this kind you know about, **is not the only tool that provides transpiling** and pollyfilling. I don't know why you are bringing that up. Also it's builtin right into the tsc compiler, see my point above about 1st party tooling.

> Minifying does not make things faster to load. Extensions are loaded from your computer, so the bottleneck only exists for those who use old computers.

Actually, a shorter code will take less space in memory, taking less time to be loaded, and less time being parsed, since shorter. This can improve the parser and lexer's performances. Also, some minifiers also add some optimisations to the code, so that the step of optimizing the bytecode can be skipped / requires less work for the JIT compiler.

> In terms of class and inheritance, vanilla features are just enough. Unless you missed ES6 class syntax, of course.

Keep in mind classes and inheritance were first added to TS long ago before being added to JS. **Newer features are already available**, like generics, assignments operators, etc. And those only the ones I can thing of right now. **Some of the can be very useful, and can reduce code complexity and improve readability.** Who wouldn't want a better experience?

> You can't make use of TypeScript if you did not provide type information. TypeScript requires YOU to define types, associate them and correctly describe them in all of the function definitions. If you just rename the files to .ts you just complicate stuff without any profits.

I mentionned that, explaining it doesn't requires much changes just to get started. The project is already good to go and changing the file extension is only the first step. Of course no one would only do that and call it a day. Did you actually thing someone would only do that?

> Based on your misunderstandings on how TypeScript works, I am going to say no.

I don't really get your point here. Are you the owner of the org? Or maybe are you the creator of SA? Why would you say no? And about what?

**Comment by Hans5958** [comment]
I'm going to take one from a lot of the points since it has been discussed before.

> it can generate typings files easily, to have an easier time using the API

To be honest, this is the one of the advantages that we're looking for. There's someone who wanted to use autocomplete on the API, and there's another one who wanted to make the docs of it. Having TS would solve this. I'm not sure how to solve this with only JS.

**Reply by WorldLanguages** [reply]
JSDoc likely the only viable solution without using TS.

**Reply by Ascor8522** [reply]
Yeah, I talked about that the other day when chatting with WorldLanguages, but it can get messy really fast. Especially when having big objects, with lots of attributes and methods. Also, not that you can only annotate the types of the parameters and the return type. Adding annotations for variables is much harder / impossible. That's why some interface, (abstract) class or type definition would be very welcome.

It is still possible to write those typings by hand so they can be used by the mainstream code editors, but that defeats the point of using JS, since those files can be generated automatically when compiling TS code...

**Reply by Hans5958** [reply]
> JSDoc likely the only viable solution without using TS.

Not the autocomplete.

**Reply by WorldLanguages** [reply]
> Not the autocomplete.

Why not? Visual Studio Code handles autocomplete with JSDoc just fine (as long as you "import" the types)

**Reply by Hans5958** [reply]
> > Not the autocomplete.
> 
> Why not? Visual Studio Code handles autocomplete with JSDoc just fine (as long as you "import" the types)

Well, that's probably one of the unconventional ways to do it, but I guess that'll do the job.

**Reply by WorldLanguages** [reply]
Yeah we got autocomplete working on Visual Studio Code with "TypeScript JSDoc" recently, you simply need to add a one line comment at the start of the file, above the default exported function in the module.

**Comment by cobaltt7** [comment]
i know nothing about typescript and if we change to typescript I'll try writing some js and people will say, "don't do that, use this newer/better typescript feature!" but I wouldn't know how and wouldn't be able to contribute anymore

**Reply by Ascor8522** [reply]
Don't worry, the goal of open source is to make code better and learn from others.
Don't take it too hard if someone at some point rewrites what you wrote.
The goal is not to blame you, just to make the code better.

Also newer features are quite groundbreaking, and have very specific use-cases.
Most of the time, plain ol' JavaScript is the best and the only way to go.

You might also read the TS handbook. It doesn't take much time, since they have a version from people coming from JS, and it's always nice to learn a couple of things here and there.

**Reply by Ascor8522** [reply]
But yeah, I know learning a new language sometimes might be overwhelming. It's normal.

**Comment by apple502j** [comment]
I am not sure what kinds of benefits we get by using TypeScript.

- Minifying is not an option (per store policies)
- No polyfills required
- Not sure what features we need - we probably don't need generics (unless we call `Array.prototype.map.call(document.querySelectorAll(), (item) => something(item))` as "generics". It's still possible in vanilla though), we already have assignment operators in vanilla and nobody needs `&&=`, etc

Statistics:

- v1.7.0: of all 39 bugs fixed, I don't think there are more than 2 bugs that TypeScript could have prevented.
- v1.5.0: #954 is probably the only type-related bug afaik.

ScratchAddons is not very big, we can manually test everything.

**Reply by WorldLanguages** [reply]
> Minifying is not an option (per store policies)

This is technically not true, minifying is OK. Mozilla would also ask for the actual source code.
Minifying is useless tho, since translations will be most of the extension bundle anyway.
